File: 1.0/src/cz/mg/vulkan/PFNvkDebugUtilsMessengerCallbackEXT.java
package cz.mg.vulkan;

public class PFNvkDebugUtilsMessengerCallbackEXT extends VkFunctionPointer {
    public PFNvkDebugUtilsMessengerCallbackEXT() {
    }

    protected PFNvkDebugUtilsMessengerCallbackEXT(VkMemory vkmemory) {
        super(vkmemory);
    }

    protected PFNvkDebugUtilsMessengerCallbackEXT(VkMemory vkmemory, long vkaddress) {
        super(vkmemory, vkaddress);
    }



    public PFNvkDebugUtilsMessengerCallbackEXT(long value) {
        setValue(value);
    }

    public PFNvkDebugUtilsMessengerCallbackEXT(VkInstance instance) {
        super(instance, new VkString("vkDebugUtilsMessengerCallbackEXT"));
    }

    public int call(int messageSeverity, int messageType, VkDebugUtilsMessengerCallbackDataEXT pCallbackData, VkObject pUserData){
        return callNative(getValue(), messageSeverity, messageType, pCallbackData != null ? pCallbackData.getVkAddress() : VkPointer.NULL, pUserData != null ? pUserData.getVkAddress() : VkPointer.NULL);
    }

    protected static native int callNative(long vkaddress, int messageSeverity, int messageType, long pCallbackData, long pUserData);

}

File: 1.0/src/c/cz/mg/vulkan/PFN_vkDebugUtilsMessengerCallbackEXT.c
#include <jni.h>
#include <stdlib.h>
#include <string.h>
#include <vulkan/vulkan.h>

jlong jniPointerToLong(const void* p);
void* jniLongToPointer(jlong l);
jlong jniFunctionPointerToLong(PFN_vkVoidFunction p);
PFN_vkVoidFunction jniLongToFunctionPointer(jlong l);
void jniThrowException(JNIEnv* env, const char* message);

void Java_cz_mg_vulkan_PFNvkDebugUtilsMessengerCallbackEXT_callNative(JNIEnv* env, jclass jc, jlong address, jlong messageSeverity, jlong messageType, jlong pCallbackData, jlong pUserData, jlong rval){
    (void)env;
    (void)jc;
    PFN_vkDebugUtilsMessengerCallbackEXT f = (PFN_vkDebugUtilsMessengerCallbackEXT)jniLongToFunctionPointer(address);
    VkBool32* rvalAddress = jniLongToPointer(rval);
    *rvalAddress = f(
        *((VkDebugUtilsMessageSeverityFlagBitsEXT*)jniLongToPointer(messageSeverity)),
        *((VkDebugUtilsMessageTypeFlagsEXT*)jniLongToPointer(messageType)),
        ((VkDebugUtilsMessengerCallbackDataEXT*)jniLongToPointer(pCallbackData)),
        ((void*)jniLongToPointer(pUserData))
    );
}


jint Java_cz_mg_vulkan_PFNvkDebugUtilsMessengerCallbackEXT_callSimplifiedNative(JNIEnv* env, jclass jc, jlong address, jint messageSeverity, jint messageType, jlong pCallbackData, jlong pUserData){
    (void)env;
    (void)jc;
    PFN_vkDebugUtilsMessengerCallbackEXT f = (PFN_vkDebugUtilsMessengerCallbackEXT)jniLongToFunctionPointer(address);
    return f(
        (VkDebugUtilsMessageSeverityFlagBitsEXT)messageSeverity,
        (VkDebugUtilsMessageTypeFlagsEXT)messageType,
        ((VkDebugUtilsMessengerCallbackDataEXT*)jniLongToPointer(pCallbackData)),
        ((void*)jniLongToPointer(pUserData))
    );
}

